doc/manual/calling-c-and-fortran-code.rst

```diff
@@ -743,14 +736,90 @@ any expression, such as ``&0`` or ``&f(x)``.
 When a scalar value is passed with ``&`` as an argument of type
 ``Ptr{T}``, the value will first be converted to type ``T``.
 
+Some Examples of C Wrappers
+---------------------------
+
+Here is a simple example of a C wrapper that returns a ``Ptr`` type::
+
+    type gsl_permutation
+    end
+
+    # The corresponding C signature is
+    #     gsl_permutation * gsl_permutation_alloc (size_t n);
+    function permutation_alloc(n::Integer)
+        output_ptr = ccall( (:gsl_permutation_alloc, :libgsl),
+            Ptr{gsl_permutation}, (Csize_t,), n )
+        if output_ptr==C_NULL #Could not allocate memory
+            throw(MemoryError())
+        end
+        return output_ptr
+    end
+
+The `GNU Scientific Library <https://www.gnu.org/software/gsl/>`_ (here assumed
+to be accessible through ``:libgsl``) defines an opaque pointer,
+``gsl_permutation *``, as the return type of the C function
+``gsl_permutation_alloc()``. As user code never has to look inside the
+``gsl_permutation`` struct, the corresponding Julia wrapper simply needs a new
+type declaration, ``gsl_permutation``, that has no internal fields and whose
+sole purpose is to be placed in the type parameter of a ``Ptr`` type.  The
+return type of the ``ccall`` is declared as ``Ptr{gsl_permutation}``, since the
+memory allocated and pointed to by ``output_ptr`` is controlled by C (and not
+Julia).
+
+The input ``n`` is passed by value, and so the function's input signature is
+simply declared as ``(Csize_t,)`` without any ``Ref`` or ``Ptr`` necessary.
+(If the wrapper was calling a Fortran function instead, the corresponding
+function input signature should instead be ``(Ref{Csize_t},)``, since Fortran
+variables are passed by reference.) Furthermore, ``n`` can be any type that is
+convertable to a ``Csize_t`` integer; the ``ccall`` implicitly calls
+:func:`Base.cconvert(Csize_t, n) <cconvert>`.
+
+
+Here is a second example wrapping the corresponding destructor::
+
+    # The corresponding C signature is
+    #     void gsl_permutation_free (gsl_permutation * p);
+    function permutation_free(p::Ref{gsl_permutation})
+        ccall( (:gsl_permutation_free, :libgsl), Void,
+            (Ref{gsl_permutation},), p )
+    end
+
+Here, the input ``p`` is declared to be of type ``Ref{gsl_permutation}``.
+``p`` may be of type ``Ptr{gsl_permutation}``, as may be returned by
+``permutation_alloc()``, but in principle could also be constructed from memory
+managed by Julia.
+
+
+Here is a third example passing Julia arrays::
+
+    # The corresponding C signature is
+    #    int gsl_sf_bessel_Jn_array (int nmin, int nmax, double x,
+    #                                double result_array[])
+    function sf_bessel_Jn_array(nmin::Integer, nmax::Integer, x::Real)
+        if nmax<nmin throw(DomainError()) end
+        result_array = Array(Cdouble, nmax-nmin+1)
+        errno = ccall( (:gsl_sf_bessel_Jn_array, :libgsl), Cint, (Cint, Cint,
```

Comment:
Calling this `errno` is a bit confusing as it's different from the eponymous function. `err` or `errcode` would be better.